ethereum/src/ERC20Manager.sol

```diff
@@ -0,0 +1,123 @@
+
+import {IERC20Manager} from "./interfaces/IERC20Manager.sol";
+import {IMessageQueueReceiver} from "./interfaces/IMessageQueue.sol";
+import {ERC20GearSupply} from "./ERC20GearSupply.sol";
+import {BridgingPayment} from "./BridgingPayment.sol";
+
+contract ERC20Manager is IERC20Manager, IMessageQueueReceiver {
+    using SafeERC20 for IERC20;
+
+    address immutable MESSAGE_QUEUE_ADDRESS;
+    bytes32 immutable VFT_MANAGER_ADDRESS;
+
+    mapping(address => SupplyType) tokenSupplyType;
+
+    constructor(address message_queue, bytes32 vft_manager) {
+        MESSAGE_QUEUE_ADDRESS = message_queue;
+        VFT_MANAGER_ADDRESS = vft_manager;
+    }
+
+    /** @dev Request token bridging. When the bridging is requested tokens are burned/locked (based on the type of supply)
+     * from account that've sent transaction and `BridgingRequested` event is emitted that later can be verified
+     * on other side of bridge.
+     *
+     * @param token token address to transfer over bridge
+     * @param amount quantity of tokens to transfer over bridge
+     * @param to destination of transfer on gear
+     */
+    function requestBridging(address token, uint256 amount, bytes32 to) public {
+        SupplyType supply_type = tokenSupplyType[token];
+
+        if (supply_type == SupplyType.Gear) {
+            ERC20GearSupply(token).burnFrom(tx.origin, amount);
+        } else {
+            if (supply_type == SupplyType.Unknown) {
+                tokenSupplyType[token] = SupplyType.Ethereum;
+            }
+
+            IERC20(token).safeTransferFrom(tx.origin, address(this), amount);
+        }
+
+        emit BridgingRequested(tx.origin, to, token, amount);
+    }
+
+    /** @dev Accept bridging request made on other side of bridge.
+     * This request must be sent by `MessageQueue` only. When such a request is accepted, tokens
+     * are minted/unlocked to the corresponding account address, specified in `payload`.
+     *
+     * Expected `payload` consisits of these:
+     *  - `receiver` - account to mint tokens to
+     *  - `token` - token to mint
+     *  - `amount` - amount of tokens to mint
+     *
+     * Expected sender should be `vft-manager` program on gear.
+     *
+     * @param sender sender of message on the gear side.
+     * @param payload payload of the message.
+     */
+    function processVaraMessage(
+        bytes32 sender,
+        bytes calldata payload
+    ) external returns (bool) {
+        if (msg.sender != MESSAGE_QUEUE_ADDRESS) {
+            revert NotAuthorized();
+        }
+        if (payload.length != 20 + 20 + 32) {
+            revert BadArguments();
+        }
+        if (sender != VFT_MANAGER_ADDRESS) {
+            revert BadVftManagerAddress();
+        }
+
+        address receiver = address(bytes20(payload[0:20]));
+        address token = address(bytes20(payload[20:40]));
+        uint256 amount = uint256(bytes32(payload[40:]));
+
+        SupplyType supply_type = tokenSupplyType[token];
```

Comment:
It appears that the `tokenSupplyType` mapping defaults to `SupplyType.Unknown` for uninitialized tokens, meaning that the token type is only set upon the first call to `requestBridging` or `processVaraMessage`.
If the token type in this function is `SupplyType.Gear`, should this value already be set in the `processVaraMessage` function?
Perhaps it would be better to define token types in advance through an explicit setup function.